# Livro para consulta:
- https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html
- https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html
    

# 1. Importando bibliotecas <a name="import"></a>

<div style="text-align: right"
     
[Voltar ao índice](#Contents)

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

%matplotlib inline

# 2. Carregando o dataframe SINASC <a name="read"></a>
<div style="text-align: right"
     
[Voltar ao índice](#Contents)

In [3]:
sinasc_raw = pd.read_csv('SINASC_RO_2019.csv')
sinasc_raw.head()

ORIGEM   CODESTAB  CODMUNNASC  LOCNASC  IDADEMAE  ESTCIVMAE  \
0       1  2679477.0      110001        1        19        5.0   
1       1  2679477.0      110001        1        29        2.0   
2       1  2679477.0      110001        1        37        9.0   
3       1  2516500.0      110001        1        30        5.0   
4       1  2516500.0      110001        1        30        2.0   

            ESCMAE  CODOCUPMAE  QTDFILVIVO  QTDFILMORT  ...  KOTELCHUCK  \
0      8 a 11 anos         NaN         0.0         0.0  ...         5.0   
1      8 a 11 anos    999992.0         1.0         0.0  ...         5.0   
2      8 a 11 anos    513205.0         2.0         0.0  ...         5.0   
3  12 anos ou mais    231205.0         0.0         0.0  ...         4.0   
4      8 a 11 anos    999992.0         1.0         0.0  ...         5.0   

  CONTADOR munResStatus munResTipo               munResNome  munResUf  \
0      1.0        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   
1      2.0        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   
2      3.0        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   
3      4.0        ATIVO      MUNIC  Alto Alegre dos Parecis  Rondônia   
4      5.0        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   

   munResLat munResLon  munResAlt  munResArea  
0  -11.93554 -61.99982      338.0    7067.025  
1  -11.93554 -61.99982      338.0    7067.025  
2  -11.93554 -61.99982      338.0    7067.025  
3  -12.13178 -61.85308      397.0    3958.273  
4  -11.93554 -61.99982      338.0    7067.025  

[5 rows x 69 columns]

In [4]:
print('Quantidade de linhas duplicadas:',
      sinasc_raw.duplicated().sum())

Quantidade de linhas duplicadas: 0


In [5]:
sinasc_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16064 entries, 0 to 16063
Data columns (total 69 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ORIGEM        16064 non-null  int64  
 1   CODESTAB      15999 non-null  float64
 2   CODMUNNASC    16064 non-null  int64  
 3   LOCNASC       16064 non-null  int64  
 4   IDADEMAE      16064 non-null  int64  
 5   ESTCIVMAE     15883 non-null  float64
 6   ESCMAE        15900 non-null  object 
 7   CODOCUPMAE    14596 non-null  float64
 8   QTDFILVIVO    15158 non-null  float64
 9   QTDFILMORT    14857 non-null  float64
 10  CODMUNRES     16064 non-null  int64  
 11  GESTACAO      15414 non-null  object 
 12  GRAVIDEZ      16024 non-null  object 
 13  PARTO         16041 non-null  object 
 14  CONSULTAS     16064 non-null  int64  
 15  DTNASC        16064 non-null  object 
 16  HORANASC      16055 non-null  float64
 17  SEXO          16060 non-null  object 
 18  APGAR1        16016 non-nu

# Tarefa 1

### 1. Idade media das mães e dos pais por município (coluna munResNome)


In [6]:
sinasc_raw.groupby(by='munResNome')[['IDADEMAE',
                                     'IDADEPAI']].agg({'IDADEMAE':[('Média','mean')],
                                                       'IDADEPAI':[('Média','mean')]})

IDADEMAE   IDADEPAI
                               Média      Média
munResNome                                     
Alta Floresta D'Oeste      25.924901  29.540284
Alto Alegre dos Parecis    24.967742  29.588235
Alto Paraíso               24.962441  24.250000
Alvorada D'Oeste           25.787879  31.333333
Ariquemes                  25.622461  31.333333
Buritis                    25.560748  30.516667
Cabixi                     25.228571  34.666667
Cacaulândia                25.536232  36.666667
Cacoal                     26.893473  30.894184
Campo Novo de Rondônia     24.982301  30.192308
Candeias do Jamari         25.491329  29.615385
Castanheiras               26.882353  30.333333
Cerejeiras                 27.116129  31.241935
Chupinguaia                25.535354  30.055556
Colorado do Oeste          27.855769  33.400000
Corumbiara                 24.542373  31.454545
Costa Marques              24.894040  31.240000
Cujubim                    24.584699  32.000000
Espigão D'Oeste            25.694545  30.753769
Governador Jorge Teixeira  24.203125  29.800000
Guajará-Mirim              25.495169  30.800000
Itapuã do Oeste            26.970149  44.833333
Jaru                       26.412417  31.684211
Ji-Paraná                  26.156951  30.715074
Machadinho D'Oeste         25.576803  30.942857
Ministro Andreazza         27.907895  34.030303
Mirante da Serra           25.393617  33.094340
Monte Negro                25.341317  34.720000
Município ignorado - RO    24.000000  22.000000
Nova Brasilândia D'Oeste   26.206107  30.542857
Nova Mamoré                25.461187  31.866667
Nova União                 26.720000  29.333333
Novo Horizonte do Oeste    27.521739  32.439024
Ouro Preto do Oeste        26.350427  31.222727
Parecis                    26.434783  32.333333
Pimenta Bueno              25.251445  29.699324
Pimenteiras do Oeste       25.739130  35.600000
Porto Velho                26.285947  32.584848
Presidente Médici          25.565476  30.200000
Primavera de Rondônia      27.538462  32.208333
Rio Crespo                 26.666667  28.000000
Rolim de Moura             26.812121  30.533040
Santa Luzia D'Oeste        25.982456  30.320755
Seringueiras               25.573171  32.263158
São Felipe D'Oeste         26.324324  31.294118
São Francisco do Guaporé   25.770115  31.250000
São Miguel do Guaporé      26.158416  32.278689
Teixeirópolis              25.888889  31.321429
Theobroma                  25.562500  30.833333
Urupá                      26.857143  31.491525
Vale do Anari              26.295082  32.636364
Vale do Paraíso            24.800000  30.458333
Vilhena                    26.501832  34.161290

### 2. Peso médio dos bebes por sexo que nasceram no dia do seu aniversário por faixas de escolaridade mae
Ex: Você, aluna(o), nasceu no dia 10/01, então você precisa filtrar o conjunto de dados nessa data e calcular o peso médio dos bebês de cada sexo por faixa de escolaridade da mãe.

In [7]:
sinasc_raw.DTNASC = pd.to_datetime(sinasc_raw.DTNASC)
sinasc_raw.DTNASC

0       2019-02-19
1       2019-02-21
2       2019-02-25
3       2019-03-20
4       2019-03-23
           ...    
16059   2019-08-06
16060   2019-08-09
16061   2019-08-23
16062   2019-08-23
16063   2019-08-25
Name: DTNASC, Length: 16064, dtype: datetime64[ns]

In [12]:
sinasc_day = sinasc_raw[(sinasc_raw.DTNASC.dt.day == 26) &
                                (sinasc_raw.DTNASC.dt.month == 1) &
                                (sinasc_raw.DTNASC.dt.year == 2019)]

print('Em 2019, dos {} nascimentos registrados no Estado de Rondônia, {} ocorreram no dia 26 de janeiro.'.format(
    len(sinasc_raw),
    len(sinasc_day)
))


sinasc_day = sinasc_day.groupby(by=['DTNASC',
                                                    'SEXO',
                                                    'ESCMAE'])['PESO'].mean()

sinasc_day = sinasc_day.to_frame().reindex(labels=['1 a 3 anos',
                                                                   '4 a 7 anos',
                                                                   '8 a 11 anos',
                                                                   '12 anos ou mais'],
                                                           level='ESCMAE')

sinasc_day

Em 2019, dos 16064 nascimentos registrados no Estado de Rondônia, 55 ocorreram no dia 26 de janeiro.


PESO
DTNASC     SEXO      ESCMAE                      
2019-01-26 Feminino  4 a 7 anos       3003.333333
                     8 a 11 anos      3337.736842
                     12 anos ou mais  3208.333333
           Masculino 4 a 7 anos       3335.250000
                     8 a 11 anos      3367.941176
                     12 anos ou mais  3375.000000

### 3. Qual o municipio que nasceu menos bebe em 2019?
    - qual a idade media, maxima, minima das maes nesse municipio?
    - qual a idade media, maxima, minima dos pais nesse municipio?

In [13]:
print('Municípios em que nasceram menos bebês:\n')
print(sinasc_raw.munResNome.value_counts().sort_values().head())

sinasc_raw['COUNT'] = 1
sinasc_3 = sinasc_raw.groupby(by='munResNome')[['COUNT',
                                                'IDADEMAE',
                                                'IDADEPAI']].agg(
    {'COUNT':[('Quantidade de bebês nascidos','count')],
     'IDADEMAE':[('idade média','mean'),
                 ('idade máxima',max),
                 ('idade mínima',min)],
     'IDADEPAI':[('idade média','mean'),
                 ('idade máxima',max),
                 ('idade mínima',min)]})

sinasc_3.sort_values(('COUNT', 'Quantidade de bebês nascidos'))

Municípios em que nasceram menos bebês:

munResNome
Município ignorado - RO     1
Castanheiras               17
Parecis                    23
Pimenteiras do Oeste       23
Primavera de Rondônia      26
Name: count, dtype: int64


COUNT    IDADEMAE  \
                          Quantidade de bebês nascidos idade média   
munResNome                                                           
Município ignorado - RO                              1   24.000000   
Castanheiras                                        17   26.882353   
Pimenteiras do Oeste                                23   25.739130   
Parecis                                             23   26.434783   
Primavera de Rondônia                               26   27.538462   
Cabixi                                              35   25.228571   
Teixeirópolis                                       36   25.888889   
São Felipe D'Oeste                                  37   26.324324   
Vale do Paraíso                                     45   24.800000   
Novo Horizonte do Oeste                             46   27.521739   
Theobroma                                           48   25.562500   
Rio Crespo                                          48   26.666667   
Nova União                                          50   26.720000   
Santa Luzia D'Oeste                                 57   25.982456   
Corumbiara                                          59   24.542373   
Vale do Anari                                       61   26.295082   
Governador Jorge Teixeira                           64   24.203125   
Itapuã do Oeste                                     67   26.970149   
Cacaulândia                                         69   25.536232   
Ministro Andreazza                                  76   27.907895   
Seringueiras                                        82   25.573171   
Urupá                                               84   26.857143   
Alto Alegre dos Parecis                             93   24.967742   
Mirante da Serra                                    94   25.393617   
Chupinguaia                                         99   25.535354   
Alvorada D'Oeste                                    99   25.787879   
Colorado do Oeste                                  104   27.855769   
Campo Novo de Rondônia                             113   24.982301   
Nova Brasilândia D'Oeste                           131   26.206107   
Costa Marques                                      151   24.894040   
Cerejeiras                                         155   27.116129   
Monte Negro                                        167   25.341317   
Presidente Médici                                  168   25.565476   
Candeias do Jamari                                 173   25.491329   
São Francisco do Guaporé                           174   25.770115   
Cujubim                                            183   24.584699   
São Miguel do Guaporé                              202   26.158416   
Alto Paraíso                                       213   24.962441   
Nova Mamoré                                        219   25.461187   
Alta Floresta D'Oeste                              253   25.924901   
Espigão D'Oeste                                    275   25.694545   
Machadinho D'Oeste                                 319   25.576803   
Buritis                                            321   25.560748   
Pimenta Bueno                                      346   25.251445   
Ouro Preto do Oeste                                351   26.350427   
Guajará-Mirim                                      414   25.495169   
Jaru                                               451   26.412417   
Rolim de Moura                                     495   26.812121   
Vilhena                                            819   26.501832   
Ji-Paraná                                         1115   26.156951   
Cacoal                                            1333   26.893473   
Ariquemes                                         1674   25.622461   
Porto Velho                                       4305   26.285947   

                                                       IDADEPAI               \
                          idade máxima idade mín

### 4. Qual o municipio que nasceu mais bebe no mês de março?
    - qual a quantidade de filhos vivos media, maxima, minima nesse municipio?
    - qual a idade media, maxima, minima dos pais nesse municipio?



In [14]:
sinasc_mar = sinasc_raw[sinasc_raw.DTNASC.dt.month == 3].copy()

print('O município que mais nasceu bebê no mês de março foi {} com {} nascimentos.'.format(
    sinasc_mar['munResNome'].value_counts().index[0],
    sinasc_mar['munResNome'].value_counts().values[0]))

sinasc_mar['MES'] = sinasc_mar['DTNASC'].dt.month_name()
sinasc_mar['COUNT'] = 1
sinasc_4 = sinasc_mar.groupby(by=['MES',
                                  'munResNome']
                             )[['COUNT',
                                'QTDFILVIVO',
                                'IDADEMAE',
                                'IDADEPAI']
                              ].agg({'COUNT':[('Qtd de bebês nascidos','count')],
                                     'QTDFILVIVO':[('Média','mean'),
                                                   ('Máx',max),
                                                   ('Mín',min)],
                                     'IDADEMAE':[('Média','mean'),
                                                 ('Máx',max),
                                                 ('Mín',min),
                                                 ('Amplitude',lambda x: x.max()-x.min())],
                                     'IDADEPAI':[('Média','mean'),
                                                 ('Máx',max),
                                                 ('Mín',min),
                                                 ('Amplitude',lambda x: x.max()-x.min())]
                                    })

sinasc_4['Diferença da média IDADEPAI-IDADEMAE'] = sinasc_4[('IDADEPAI','Média')]-sinasc_4[('IDADEMAE','Média')]
sinasc_4.sort_values(('COUNT', 'Qtd de bebês nascidos'),
                     ascending=False
                    )

O município que mais nasceu bebê no mês de março foi Porto Velho com 744 nascimentos.


COUNT QTDFILVIVO            \
                                Qtd de bebês nascidos      Média  Máx  Mín   
MES   munResNome                                                             
March Porto Velho                                 744   1.295056  9.0  0.0   
      Ji-Paraná                                   188   0.828877  4.0  0.0   
      Vilhena                                     148   0.979592  5.0  0.0   
      Ariquemes                                   141   1.064748  4.0  0.0   
      Cacoal                                      133   0.969697  7.0  0.0   
      Jaru                                         80   1.025641  5.0  0.0   
      Guajará-Mirim                                75   1.534247  8.0  0.0   
      Rolim de Moura                               69   0.724638  5.0  0.0   
      Ouro Preto do Oeste                          67   1.014925  5.0  0.0   
      Pimenta Bueno                                62   0.836066  4.0  0.0   
      Buritis                                      59   1.421053  4.0  0.0   
      Nova Mamoré                                  53   1.360000  6.0  0.0   
      Espigão D'Oeste                              44   1.386364  6.0  0.0   
      Machadinho D'Oeste                           44   1.000000  6.0  0.0   
      São Miguel do Guaporé                        41   1.000000  4.0  0.0   
      Candeias do Jamari                           33   1.250000  4.0  0.0   
      São Francisco do Guaporé                     31   1.161290  6.0  0.0   
      Alta Floresta D'Oeste                        31   0.903226  4.0  0.0   
      Presidente Médici                            29   0.724138  3.0  0.0   
      Costa Marques                                24   1.208333  4.0  0.0   
      Alvorada D'Oeste                             22   1.000000  3.0  0.0   
      Alto Paraíso                                 22   1.285714  3.0  0.0   
      Monte Negro                                  20   0.700000  3.0  0.0   
      Nova Brasilândia D'Oeste                     20   0.947368  3.0  0.0   
      Colorado do Oeste                            19   1.000000  4.0  0.0   
      Seringueiras                                 19   1.000000  2.0  0.0   
      Chupinguaia                                  18   1.055556  4.0  0.0   
      Cujubim                                      17   1.294118  5.0  0.0   
      Cerejeiras                                   16   0.875000  2.0  0.0   
      Urupá                                        14   0.928571  2.0  0.0   
      Vale do Paraíso                              13   0.583333  2.0  0.0   
      Campo Novo de Rondônia                       13   1.285714  2.0  0.0   
      Corumbiara                                   12   0.909091  3.0  0.0   
      Vale do Anari                                11   0.818182  2.0  0.0   
      Alto Alegre dos Parecis                      10   0.888889  2.0  0.0   
      Governador Jorge Teixeira                    10   0.555556  2.0  0.0   
      Cacaulândia                                  10   1.200000  4.0  0.0   
      Rio Crespo                                    9   1.000000  3.0  0.0   
      Mirante da Serra                              9   1.000000  2.0  0.0   
      Theobroma                                     8   1.125000  2.0  0.0   
      Itapuã do Oeste                               8   1.285714  3.0  0.0   
      Santa Luzia D'Oeste                           7   0.428571  2.0  0.0   
      Nova União                                    7   0.571429  1.0  0.0   
      Cabixi                                        6   0.833333  3.0  0.0   
      Parecis                                       6   1.833333  5.0  1.0   
      Ministro Andreazza                            6   1.166667  3.0  0.0   
      Pimenteiras do Oeste                          6   1.000000  2.0  0.0   
      Primavera de Rondônia                         5   1.200000  2.0  0.0   
      São Felipe D'Oeste                            5   0.000000  0.0 

In [18]:
print('Mediana de todos os municípios no mês de março:')
round(sinasc_4.groupby(by='MES').median())

Mediana de todos os municípios no mês de março:


COUNT QTDFILVIVO           IDADEMAE              \
      Qtd de bebês nascidos      Média  Máx  Mín    Média   Máx   Mín   
MES                                                                     
March                  18.0        1.0  3.0  0.0     26.0  37.0  16.0   

                IDADEPAI                        \
      Amplitude    Média   Máx   Mín Amplitude   
MES                                              
March      21.0     31.0  40.0  24.0      17.0   

      Diferença da média IDADEPAI-IDADEMAE  
                                            
MES                                         
March                                  5.0

### Analise as respostas encontradas, tire algum insight delas, conte pra gente algo encontrado nos dados. Algo que você julgue relevante e novo pra você.

Com base nos resultados apresentados, podemos observar que:

- A média de filhos vivos se mantém em torno de 1, independentemente da quantidade de bebês nascidos, em todos os municípios do estado.
- A média de idade dos pais é geralmente maior do que a das mães na maioria dos municípios.
- Embora o município de Ariquemes seja um dos cinco com maior número de bebês nascidos, com um total de 141 nascimentos, a variação de idade entre os pais é uma das menores, com uma diferença de apenas 1 ano, de 28 a 29 anos.